ethcore/private-tx/src/private_transactions.rs

```diff
@@ -15,128 +15,238 @@
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
 use std::sync::Arc;
-use std::collections::{HashMap, HashSet};
+use std::cmp;
+use std::collections::HashMap;
+use std::collections::hash_map::Entry;
 
 use bytes::Bytes;
 use ethcore_miner::pool;
 use ethereum_types::{H256, U256, Address};
+use heapsize::HeapSizeOf;
 use ethkey::Signature;
+use messages::PrivateTransaction;
+use parking_lot::RwLock;
 use transaction::{UnverifiedTransaction, SignedTransaction};
-
+use txpool;
+use txpool::{VerifiedTransaction, Verifier};
 use error::{Error, ErrorKind};
 
+type Pool = txpool::Pool<VerifiedPrivateTransaction, PrivateScorying>;
+
 /// Maximum length for private transactions queues.
 const MAX_QUEUE_LEN: usize = 8312;
+/// Transaction with the same (sender, nonce) can be replaced only if
+/// `new_gas_price > old_gas_price + old_gas_price >> SHIFT`
+const GAS_PRICE_BUMP_SHIFT: usize = 3; // 2 = 25%, 3 = 12.5%, 4 = 6.25%
 
 /// Desriptor for private transaction stored in queue for verification
-#[derive(Default, Debug, Clone, PartialEq, Eq)]
-pub struct PrivateTransactionDesc {
-	/// Hash of the private transaction
-	pub private_hash: H256,
-	/// Contract's address used in private transaction
-	pub contract: Address,
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct VerifiedPrivateTransaction {
+	/// Original private transaction
+	pub private_transaction: PrivateTransaction,
 	/// Address that should be used for verification
-	pub validator_account: Address,
+	pub validator_account: Option<Address>,
+	/// Resulted verified
+	pub transaction: SignedTransaction,
+	/// Original transaction's hash
+	pub transaction_hash: H256,
+	/// Original transaction's sender
+	pub transaction_sender: Address,
+}
+
+impl txpool::VerifiedTransaction for VerifiedPrivateTransaction {
+	type Hash = H256;
+	type Sender = Address;
+
+	fn hash(&self) -> &H256 {
+		&self.transaction_hash
+	}
+
+	fn mem_usage(&self) -> usize {
+		self.transaction.heap_size_of_children()
+	}
+
+	fn sender(&self) -> &Address {
+		&self.transaction_sender
+	}
+}
+
+#[derive(Debug)]
+pub struct PrivateScorying;
+
+impl txpool::Scoring<VerifiedPrivateTransaction> for PrivateScorying {
```

Comment:
@tomusdrw I reworked my queue's implementation in order to use txpool directly for storing transaction in it. The main problem, that I'm facing, is scoring implementation. Actually I'd like to fully re-use miner's scoring, but I don't know, how to do it, because it's bound with miner's VerifiedTransaction trait.

Reply:
So as result I had to copy-past scoring code from miner.

Reply:
Let's make the two (`Scoring` and `Ready`) generic then.